Add a pass to collect dropped variable statistics

rastogishubham opened the pull request:
This patch is inspired by @Snowy1803 excellent work in swift and the patch: https://github.com/swiftlang/swift/pull/73334/files

Add an instrumentation pass to llvm to collect dropped debug information variable statistics for every Function-level and Module-level IR pass.

This patch creates adds the class DroppedVariableStats which iterates over every DbgRecord in a function or module before and after an optimization pass and counts the number of variables who's debug information has been dropped due to that pass, then prints that output to stdout in a csv format.

I ran this patch on optdriver.cpp with the command line:

```
bin/clang -g -O2 -mllvm -dropped-variable-stats -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/shubham/Development/llvm-project/build_ninja/tools/opt -I/Users/shubham/Development/llvm-project/llvm/tools/opt -I/Users/shubham/Development/llvm-project/build_ninja/include -I/Users/shubham/Development/llvm-project/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -std=c++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT tools/opt/CMakeFiles/LLVMOptDriver.dir/optdriver.cpp.o -MF tools/opt/CMakeFiles/LLVMOptDriver.dir/optdriver.cpp.o.d -o tools/opt/CMakeFiles/LLVMOptDriver.dir/optdriver.cpp.o -c /Users/shubham/Development/llvm-project/llvm/tools/opt/optdriver.cpp
```

And I see the result:

Pass Name, Dropped Variables
'SCCPPass', 6
'InstCombinePass', 78)
'GlobalOptPass', 84
'SimplifyCFGPass', 131
'JumpThreadingPass', 385
'ModuleToFunctionPassAdaptor', 4053
'SROAPass', 4128
'ADCEPass', 5435
'PassManager<Function>', 9032

adrian-prantl reviewed (commented):
Looks useful; in the current form the patch needs comments and a test!

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -559,6 +559,35 @@ class DotCfgChangeReporter : public ChangeReporter<IRDataT<DCData>> {
+
+private:
+  using VarID =
+      std::tuple<const DILocalScope *, llvm::StringRef, unsigned, unsigned>;
```
This looks like it will conflate inlined instances of the same variable. Don't we also need an inlinedAt scope pointer?

adrian-prantl replied:
Also, maybe this should be a struct.

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -559,6 +559,35 @@ class DotCfgChangeReporter : public ChangeReporter<IRDataT<DCData>> {
+  using VarID =
+      std::tuple<const DILocalScope *, llvm::StringRef, unsigned, unsigned>;
+
+  SmallVector<std::pair<llvm::DenseSet<VarID>, std::string>>
```
I think the PassIDs are constant strings, so storing a StringRef should be sufficient.

adrian-prantl commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2441,6 +2447,84 @@ void DotCfgChangeReporter::registerCallbacks(
+    return;
+  }
+  for (auto Var : DebugVariablesBefore.back().first)
+    if (!DebugVariablesAfter.back().first.contains(Var))
```
Is this doing a linear search through the SmallVector? How about sorting the vector and using std::lower_bound here?

felipepiovezan replied:
It's as `DenseSet`

felipepiovezan commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2441,6 +2447,84 @@ void DotCfgChangeReporter::registerCallbacks(
+    for (const Instruction &I : BB) {
+      for (DbgRecord &DR : I.getDbgRecordRange()) {
+        if (auto *Dbg = dyn_cast<DbgVariableRecord>(&DR)) {
+          llvm::StringRef UniqueName = Dbg->getVariable()->getName();
```
we don't generally use `llvm::` inside the llvm subdirectory

felipepiovezan commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2441,6 +2447,84 @@ void DotCfgChangeReporter::registerCallbacks(
+
+void DroppedVariableStats::runAfterPass(StringRef PassID, Any IR,
+                                        const PreservedAnalyses &PA) {
+  assert(DebugVariablesBefore.back().second ==
```
Let's give names to things instead of `.back()` everywhere:

```
auto &[VarIDs, PassName] = + (Before ? DebugVariablesBefore : DebugVariablesAfter).back();
```

felipepiovezan commented:
I echo Adrian's comments about tests, though I am unsure how to do it.
You can probably construct a simple pipeline through `opt`, containing for example two passes, with some hand-crafted IR that will be certain to drop variables.
For example, if you have a function like this:

```
define @foo(i32 %arg) {
   %storage = alloca i32
   store i32, i32 %arg, ptr %storage
   dbg_declare(%storage, !somevar)
}
```

This will easily be optimized away / dropped

rastogishubham commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2441,6 +2447,84 @@ void DotCfgChangeReporter::registerCallbacks(
+void DroppedVariableStats::runOnFunction(const Function *F, bool Before) {
+  auto &DebugVariables = Before ? DebugVariablesBefore : DebugVariablesAfter;
+  for (auto &BB : *F) {
+    for (const Instruction &I : BB) {
```
:O

rastogishubham commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2441,6 +2447,84 @@ void DotCfgChangeReporter::registerCallbacks(
+}
+
+void DroppedVariableStats::runOnFunction(const Function *F, bool Before) {
+  auto &DebugVariables = Before ? DebugVariablesBefore : DebugVariablesAfter;
```
Ah much cleaner, thanks

rastogishubham commented:
The dropped statistics for the updated patch are:

AlwaysInlinerPass, 2
SCCPPass, 6
InstCombinePass, 19
EliminateAvailableExternallyPass, 64
SimplifyCFGPass, 157
GlobalOptPass, 192
JumpThreadingPass, 439
ADCEPass, 5536

adrian-prantl commented:
> The dropped statistics for the updated patch are:
> 
> AlwaysInlinerPass, 2 

Inlining shouldn't really be dropping any variables — how do these happen / do they make sense?

jryans commented:
Looks quite helpful, thanks for working on this! 😄 

Does this report variables that have been dropped entirely, or also ones which have lost partial coverage...? I can't quite tell from the code, as you do appear to track some variable line info...

adrian-prantl commented:
> Looks quite helpful, thanks for working on this! 😄
> 
> Does this report variables that have been dropped entirely, or also ones which have lost partial coverage...? I can't quite tell from the code, as you do appear to track some variable line info...

Currently only variables that are dropped entirely, it doesn't track fragments yet.

jmorse reviewed (commented):
LGTM, although I'd suggest you want a test to check that nothing is printed when nothing is dropped (such as when the verifier runs), purely as a sanity-check. Thanks for working on this!

What's motivated using the line-number to identify a variable, instead of using the DILocalVariable directly? There's the `DebugVariable` class out there that exists for variable-identity purposes, it even has fragment-handling built in.

In this implementation "dropped" is meaning "all the records of its location disappear" as opposed to "it's completely optimised away and all the records have an undef/poison value". I think sometimes we use the terms interchangeably, the former seems like the most useful in this context though.

rastogishubham commented:
I have updated the patch to make the dropped variable statistics more accurate. We do not want to count a variable as dropped if the function that contained the dbg_val was eliminated due to dead code elimination. Do this by checking if the `InlinedAt` for a debug variable is in the `SubprogramToFunctionMap`, only if it exists, find a real instruction in the function obtained by the `SubprogramToFunctionMap` that has the same scope as the debug variable, if such an instruction exists, the debug variable has been dropped.

The new numbers for optdriver.cpp are:

SCCPPass: 48
InstCombinePass: 561
SimplifyCFGPass: 1306
JumpThreadingPass: 4087
ADCEPass: 24992

rastogishubham commented:
@jmorse Thanks for the feedback! I added a testcase to make sure we don't report dropped variables in a pass like the verifier pass.

I also didn't think of using the DILocalVarible, it was an oversight, I have updated the patch with your feedback!

rastogishubham commented:
@adrian-prantl @jmorse @jryans just pinging again to see if there were any other changes or is this patch good to go in

adrian-prantl commented on llvm/test/Other/dropped-var-stats.ll:
```diff
@@ -0,0 +1,30 @@
+    #dbg_value(i32 %y, !15, !DIExpression(), !23)
+  %add = add nsw i32 %y, 2,!dbg !25
+  ret i32 %add,!dbg !26
+}
```
You're missing tests for the inline case and/or a lexical scope with a dbg_value that has all of its (childrens') instructions eliminated.

rastogishubham committed:
Gather dropped debug info variable statistics.

This patch adds the class DroppedVariableStats to
StandardInstrumentations which gathers information on whether debug
information was dropped by an optimization pass in llvm. This runs on
every Function-level and Module-level IR pass.

rastogishubham commented:
This patch has been heavily reworked to remove false positives from dead code elimination passes. Now, debug information is only considered dropped if a `#dbg_value` doesn't exist in the `Function` after a pass has run and there exists a real instruction that has the same scope as the `#dbg_value` or a child scope of the `#dbg_value` scope, and the inlinedAt of the instruction is the same as the `#dbg_value` inlinedAt or the `#dbg_value'`s inlinedAt is part of the inlinedAt chain of the instruction's inlinedAt.

When we run this pass on optdriver.cpp, we can see a much more reasonable:

'InstCombinePass', 1
'SimplifyCFGPass', 6
'JumpThreadingPass', 25

Variables being dropped.

rastogishubham commented:
@felipepiovezan @adrian-prantl @jmorse @jryans please let me know of any feedback!

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -578,6 +579,74 @@ class PrintCrashIRInstrumentation {
   static void SignalHandler(void *);
 };
 
+class DroppedVariableStats {
```
Can you add a Doxygen comment for this?

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -578,6 +579,74 @@ class PrintCrashIRInstrumentation {
+  SmallVector<DenseMap<StringRef, DenseSet<VarID>>> DebugVariablesBefore;
+  /// A stack of DenseMaps, which map the name of an llvm::Function to a
+  /// DenseSet of VarIDs after an optimization pass has run.
+  SmallVector<DenseMap<StringRef, DenseSet<VarID>>> DebugVariablesAfter;
```
We might be able to get away with using the Function * here?

adrian-prantl replied:
It seems unnecessary to have two separate identical stacks here, maybe have one stack with a struct containing the other sets?

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -578,6 +579,74 @@ class PrintCrashIRInstrumentation {
+  /// Iterate over all Functions in a Module and report any dropped debug
+  /// information. Will call calculateDropppedVarStatsOnFunction on every
+  /// Function.
+  void calculateDropppedVarStatsOnModule(const Module *M, StringRef PassID,
```
```suggestion
  void calculateDroppedVarStatsOnModule(const Module *M, StringRef PassID,
```

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -578,6 +579,74 @@ class PrintCrashIRInstrumentation {
+                                         std::string PassLevel);
+  /// Iterate over all Instructions in a Function and report any dropped debug
+  /// information.
+  void calculateDropppedVarStatsOnFunction(const Function *F, StringRef PassID,
```
```suggestion
  void calculateDroppedVarStatsOnFunction(const Function *F, StringRef PassID,
```

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,586 @@
+      "CSK_MD5, checksum: \"719364c4b07176af8515cac6bd21008c\")\n"
+      "!2 = !{i32 7, !\"Dwarf Version\", i32 5}\n"
+      "!3 = !{i32 2, !\"Debug Info Version\", i32 3}\n"
+      "!4 = !{i32 1, !\"wchar_size\", i32 4}\n"
```
I don't think you need any of these

rastogishubham replied:
So weirdly enough, the DWARF version is the one that is needed, but everything else is removable

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,586 @@
+      "runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, "
+      "nameTableKind: Apple, sysroot: \"/\")\n"
+      "!1 = !DIFile(filename: \"/tmp/code.cpp\", directory: "
+      "\"/Users/shubham/Development/llvm-project/build_ninja\", checksumkind: "
```
just use `/`

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,586 @@
+      "nameTableKind: Apple, sysroot: \"/\")\n"
+      "!1 = !DIFile(filename: \"/tmp/code.cpp\", directory: "
+      "\"/Users/shubham/Development/llvm-project/build_ninja\", checksumkind: "
+      "CSK_MD5, checksum: \"719364c4b07176af8515cac6bd21008c\")\n"
```
leave out the checksum

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,586 @@
+      "baff49d3a0ef8e0848a726656ebf6e7b310e5113)\"}\n"
+      "!9 = distinct !DISubprogram(name: \"foo\", linkageName: \"_Z3fooi\", "
+      "scope: !10, file: !10, line: 1, type: !11, scopeLine: 1, flags: "
+      "DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition "
```
flags are not needed either probably

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,586 @@
+  LLVMContext C;
+
+  const char *IR =
+      "; Function Attrs: mustprogress nounwind ssp uwtable(sync)\n"
```
how about using a modern C++ multiline string literal?

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,586 @@
+  Stats.runBeforePass("Test",
+                      llvm::Any(const_cast<const llvm::Module *>(M.get())));
+
+  // Remove instructions
```
Maybe spell out that this is the "pass" that is being checked.

adrian-prantl commented on llvm/unittests/IR/DroppedVariableStatsTest.cpp:
```diff
@@ -0,0 +1,446 @@
+//===- unittests/IR/DroppedVariableStatsTest.cpp - TimePassesHandler tests
```
nit: this is supposed to fit into 80 columns.

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -578,6 +579,84 @@ class PrintCrashIRInstrumentation {
+    DenseMap<const Function *, DenseSet<VarID>> DebugVariablesBefore;
+    /// Maps the name of an llvm::Function to a DenseSet of VarIDs after an
+    /// optimization pass has run.
+    DenseMap<const Function *, DenseSet<VarID>> DebugVariablesAfter;
```
And why are these two maps? I would expect them to have exactly identical keys, too?
i.e.: 

struct DebugVariables {
    DenseSet<VarID> DebugVariablesBefore;
    DenseSet<VarID> DebugVariablesAfter;
};

SmallVector<DenseMap<Function*, DebugVariables>> DebugVariablesStack;

adrian-prantl commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2514,6 +2521,181 @@ void PrintCrashIRInstrumentation::registerCallbacks(
+
+void DroppedVariableStats::runBeforePass(StringRef PassID, Any IR) {
+  DebugVariablesStack.push_back(
+      {DenseMap<const Function *, DenseSet<VarID>>(),
```
does this work too?

DebugVariablesStack.push_back({})

rastogishubham replied:
No that does not work

adrian-prantl commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2514,6 +2521,181 @@ void PrintCrashIRInstrumentation::registerCallbacks(
+  // contains the inlinedAt of the #dbg_value, if such an Instruction is found,
+  // debug information is dropped.
+  for (auto Var : DebugVariablesBeforeMap) {
+    if (!DebugVariablesAfterMap.contains(Var)) {
```
```suggestion
    if (DebugVariablesAfterMap.contains(Var)) continue;
```

adrian-prantl commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2514,6 +2521,181 @@ void PrintCrashIRInstrumentation::registerCallbacks(
+  // debug information is dropped.
+  for (auto Var : DebugVariablesBeforeMap) {
+    if (!DebugVariablesAfterMap.contains(Var)) {
+      const auto *DbgValScope = std::get<0>(Var);
```
If you can fit it within the 80 column limit, I would replace the autos here with the type names

adrian-prantl commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2514,6 +2521,181 @@ void PrintCrashIRInstrumentation::registerCallbacks(
+      const auto *DbgValScope = std::get<0>(Var);
+      for (const auto &I : instructions(F)) {
+        auto *DbgLoc = I.getDebugLoc().get();
+        if (DbgLoc) {
```
```suggestion
        if (@DbgLoc) continue
```

adrian-prantl commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -2514,6 +2521,181 @@ void PrintCrashIRInstrumentation::registerCallbacks(
+            if (isInlinedAtChildOfOrEqualTo(DbgLoc->getInlinedAt(),
+                                            InlinedAtsMap[Var])) {
+              DroppedCount++;
+              break;
```
This would benefit form a comment like:
// Found another instruction in the variable's scope, so there exists a break point at which the variable could be observed. Count it as dropped.

adrian-prantl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -578,6 +579,83 @@ class PrintCrashIRInstrumentation {
+
+private:
+  bool PassDroppedVariables = false;
+  /// VarID is a unique key that represents a #dbg_value
```
```suggestion
  /// A unique key that represents a #dbg_value.
```
